packages/editor/src/services/property.util.ts

```diff
@@ -0,0 +1,306 @@
+      let refs = _.pickBy(prev, _.flip(key => _.startsWith(key, schemaRefValue.uri)));
+      if (extractedReferences[schemaRefValue.uri]) {
+        refs = undefined;
+      }
+      if (!extractedReferences[schemaRefValue.uri]) {
+        extractedReferences[schemaRefValue.uri] = schemaRefValue.uri;
+        prev = _.omitBy(prev, value => value.uri === schemaRefValue.uri);
+      }
+      if (refs !== undefined) {
+        findReferences(parentSchema, prev, refs, extractedReferences);
+      }
+      return prev;
+    },
+    allRefs
+  );
+
+/**
+ * Calculate references for a given schema and copy definitions into the schema
+ *
+ * @param parentSchema root schema which is used to find all the schema refs
+ * @param schema current subschema without resolved references
+ * @returns JsonSchema current subschema with resolved references
+ */
+export const makeSchemaSelfContained = (parentSchema: JsonSchema,
+                                        schema: JsonSchema): JsonSchema => {
+  const schemaRefs = JsonRefs.findRefs(schema, {resolveCirculars: true});
+  const allRefs = JsonRefs.findRefs(parentSchema, {resolveCirculars: true});
+  let extractedReferences;
+  findReferences(parentSchema, allRefs, schemaRefs, extractedReferences = {});
+  const refList = _.values(extractedReferences) as string[];
+  if (!_.isEmpty(refList)) {
+    _.each(refList, ref => {
+      const propertyKey = ref.substring((ref.indexOf('/') + 1), (ref.lastIndexOf('/')));
+      const property = ref.substring((ref.lastIndexOf('/') + 1));
+      if (_.has(parentSchema, propertyKey) && _.has(parentSchema, `${propertyKey}.${property}`)) {
+        if (schema[propertyKey]) {
+          schema[propertyKey] = {
+            ...schema[propertyKey], ...{
+              [property]: parentSchema[propertyKey][property]
+            }
+          };
+        } else {
+          schema = {
+            ...schema, ...{
+              [propertyKey]: { [property]: parentSchema[propertyKey][property] }
+            }
+          };
+        }
+      }
+    });
+  }
+
+  return schema;
+};
+
+/**
+ * Create a self contained schema.
+ * @param parentSchema The schema to use for resolving
+ * @param refPath The path to resolve
+ * @return a JsonSchema that is self-contained
+ */
+export const makeSelfContainedSchema = (parentSchema: JsonSchema, refPath: string): JsonSchema => {
+  let schema = resolveSchema(parentSchema, refPath);
+  schema = deepCopy(schema);
+  if (_.isEmpty(schema.id)) {
+    schema.id = '#' + refPath;
+  }
+  schema = { ...schema, ...makeSchemaSelfContained(parentSchema, schema) };
+
+  return schema;
+};
+
+/**
+ * Calculates label for an item which is resolved in anyOf
+ * @param {string} id The schema id
+ * @returns {string}
+ */
+const calculateLabel = (id: string): string => (
+  _.startsWith(id, '#') ? id.substring(id.lastIndexOf('#') + 1) : id
+);
+
+/**
+ * Constructing container properties of a given schema. A container means
```

Comment:
Rephrase (if correct): Returns a flattened list of container properties. A property is being considered a container property if it is an array of objects, meaning that all its children are non-primitive.